### DISCUSSION TITLE
Using non Cargo specific files in the build environment

### DISCUSSION THREAD

**Discussion by Witcher01** [ipetkov/crane#296] [Q&A] [answered]
I set up a `Makefile` for a project of mine that uses Rust with standard rules, including `install`. When trying to use `make install PREFIX=$out` in the `installPhase` it errors out because there is no such rule, even though I have specified it.
Looking closer, the `Makefile` doesn't even seem to be copied into the build environment at all, all I see are `Cargo.lock`, `Cargo.toml`, `src`, and `target`.

How do I work with `Makefile`s, or any other non-cargo files?
If I can't, what's the reasoning, and what's a workaround? I don't want to *force* nix onto downstream users as I don't want anyone installing a new package manager to use my software that is not *required*.

Thanks!

flake.nix:
```nix
{
  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
    crane = {
      url = "github:ipetkov/crane";
      inputs.nixpkgs.follows = "nixpkgs";
    };
    flake-utils.url = "github:numtide/flake-utils";
    rust-overlay = {
      url = "github:oxalica/rust-overlay";
      inputs = {
        nixpkgs.follows = "nixpkgs";
        flake-utils.follows = "flake-utils";
      };
    };
  };

  outputs = { self, nixpkgs, crane, flake-utils, rust-overlay }:
    flake-utils.lib.eachDefaultSystem (system:
      let
        pkgs = import nixpkgs {
          system = system;
          overlays = [ (import rust-overlay) ];
        };

        rustToolchain = pkgs.pkgsBuildHost.rust-bin.fromRustupToolchainFile
          ./rust-toolchain.toml;
        lib = crane.lib.${system}.overrideToolchain rustToolchain;

        rss-email = lib.buildPackage {
          VERSION = "test";
          src = ./.;
          nativeBuildInputs = with pkgs; [ ];
          installCargoArtifactsMode = "use-symlink";
          installPhase = ''
            mkdir -p $out
            make install PREFIX=$out
          '';
        };
      in
      {
        packages.default = rss-email;
      }
    );
}
```

**Comment by dpc** [comment, marked as answer]
By default, if you don't take care of splitting your build into deps-only + normal (using `cargoArtifacts`), the crane lib will do it for you and both of these will run your `installPhase`. Usually such default behavior OK, but here it will not work, as in deps-only build, almost all files/code is missing and `make` needs them.

 https://github.com/ipetkov/crane/blob/master/docs/API.md#optional-attributes-1

**Reply by ipetkov** [reply]
This is exactly what's happening here! Perhaps `buildPackage` should automatically clear out `installPhase` for `buildDepsOnly` (when `cargoArtifacts` are not specified) to avoid surprises like this :thinking:

**Reply by ipetkov** [reply]
This should be improved with https://github.com/ipetkov/crane/pull/297

**Reply by Witcher01** [reply]
I see, thank you for your help and the prompt reply!

@ipetkov Thank you, too, for clarifying this so quickly!
